### PR TITLE
[mixer] Mixer IL Extern function refactoring

### PR DESCRIPTION
**Pull request by ozevren**
**What this PR does / why we need it**:
Refactors extern functions into its own library,"pkg/il/runtime".

+ It also consolidates some of the duplicate implementations.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: mixer/pkg/il/compiler/compiler_test.go [hunk 1]
```diff
@@ -15,1172 +15,50 @@
...
-			},
-		},
-	},
-}
-
 func TestCompile(t *testing.T) {
 
-	finder := descriptor.NewFinder(&globalConfig)
+	for i, test := range ilt.TestData {
+		if test.E == "" {
```
Comment (guptasu) [inline mixer/pkg/il/compiler/compiler_test.go]: Suggest : Please clarify what does this E mean

File: mixer/pkg/il/compiler/compiler_test.go [hunk 2]
```diff
@@ -15,1172 +15,50 @@
...
+			continue
+		}
 
-	for i, te := range tests {
-		t.Run(fmt.Sprintf("%d '%s'", i, te.expr), func(tt *testing.T) {
-			result, err := Compile(te.expr, finder)
+		conf := test.Conf
```
Comment (guptasu) [inline mixer/pkg/il/compiler/compiler_test.go]: Curious: is there a reason why below lines are not inside the `t.Run (...`. AFAIK, that will do nice parallelization for all tests inside the ilt.TestData
```
+		conf := test.Conf
 +		if conf == nil {
 +			conf = ilt.TestConfigs["Default"]
 +		}
 +		finder := descriptor.NewFinder(conf)
 +
 +		name := fmt.Sprintf("%d '%s'", i, test.E)
```

File: mixer/pkg/il/compiler/compiler_test.go [hunk 3]
```diff
@@ -15,1172 +15,50 @@
...
+				}
+				return
 			}
-			actual := text.WriteText(result.Program)
-			if len(te.code) > 0 {
-				if strings.TrimSpace(actual) != strings.TrimSpace(te.code) {
+
+			if test.CompileErr != "" {
+				tt.Fatalf("expected error not found: '%s'", test.CompileErr)
+				return
```
Comment (guptasu) [inline mixer/pkg/il/compiler/compiler_test.go]: probably don't need return after Fatalf

File: mixer/pkg/il/compiler/compiler_test.go [hunk 4]
```diff
@@ -1189,91 +67,27 @@ func TestCompile(t *testing.T) {
...
-			} else if v.AsInterface() != te.result {
-				tt.Fatalf("Result match failed: %+v == %+v", v.AsInterface(), te.result)
+			if !ilt.AreEqual(test.R, v.AsInterface()) {
+				tt.Fatalf("Result match failed: %+v == %+v", test.R, v.AsInterface())
 			}
 		})
 	}
 }
-
-func TestCompile_ParseError(t *testing.T) {
```
Comment (guptasu) [inline mixer/pkg/il/compiler/compiler_test.go]: Are these tests mapped in the new test data ? And the below one too.

File: mixer/pkg/il/evaluator/evaluator_test.go [hunk 5]
```diff
@@ -26,68 +27,120 @@ import (
...
-	_, err := e.Eval("ip(\"not-an-ip-addr\")", bag)
-	if err == nil {
-		t.Fatal("Was expecting an error")
+	if ebool, ok := test.R.(bool); ok {
+		abool, err := evaluator.EvalPredicate(test.E, bag)
+		if err != nil {
+			t.Errorf("Unexpected error: %v", err)
+		}
+		if abool != ebool {
+			t.Errorf("EvalString failed: '%s' != '%s'", abool, ebool)
```
Comment (guptasu) [inline mixer/pkg/il/evaluator/evaluator_test.go]: EvalPredicate failed

File: mixer/pkg/il/evaluator/evaluator_test.go [hunk 6]
```diff
@@ -158,65 +220,6 @@ func TestConcurrent(t *testing.T) {
...
-			t.Fatalf("Unexpected error: %+v", err)
-		}
-		if r != test.result {
-			t.Logf("Expression: %s", expr)
-			t.Fatalf("Result mismatch: E:%v != A:%v", test.result, r)
-		}
-	}
-}
-
-func TestEvalPredicate_Error(t *testing.T) {
```
Comment (guptasu) [inline mixer/pkg/il/evaluator/evaluator_test.go]: Is this test accounted for in the test data list ?

File: mixer/pkg/il/compiler/compiler_test.go (hunk 2 above)
Comment (ozevren) [reply mixer/pkg/il/compiler/compiler_test.go]: Short answer: Nothing particular, moved code. Long Answer: I started with the idea of writing an AST version of this as well, but decided to consolidate all AST/IL comparison in evaluator instead.

File: mixer/pkg/il/compiler/compiler_test.go (hunk 4 above)
Comment (ozevren) [reply mixer/pkg/il/compiler/compiler_test.go]: Yes, they are both captured in the test set:
https://github.com/ozevren/istio/blob/08e84650e43f30bdbb9001561b7076e0252a7b0e/mixer/pkg/il/testing/tests.go#L1444

https://github.com/ozevren/istio/blob/08e84650e43f30bdbb9001561b7076e0252a7b0e/mixer/pkg/il/testing/tests.go#L79

File: mixer/pkg/il/evaluator/evaluator_test.go (hunk 6 above)
Comment (ozevren) [reply mixer/pkg/il/evaluator/evaluator_test.go]: Good catch. The new test infrastructure is not handling this case. I've reintroduced them.

File: mixer/pkg/il/compiler/compiler_test.go (hunk 3 above)
Comment (ozevren) [reply mixer/pkg/il/compiler/compiler_test.go]: Actually, it turns out, I do. :) Without these, the tests do not work as expected.

File: mixer/pkg/il/runtime/externs.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	"time"
+
+	"istio.io/istio/mixer/pkg/il/interpreter"
+)
+
+// Externs contains the list of standard external functions used during evaluation.
+var Externs = map[string]interpreter.Extern{
+	"ip":              interpreter.ExternFromFn("ip", externIp),
+	"ip_equal":        interpreter.ExternFromFn("ip_equal", externIpEqual),
+	"timestamp":       interpreter.ExternFromFn("timestamp", externTimestamp),
```
Comment (geeknoid) [inline mixer/pkg/il/runtime/externs.go]: Will we support standard comparisons operators ((==, >, etc) on timestamps and ip addresses too? Or is that not supported by CEL?

Comment (ozevren) [reply mixer/pkg/il/runtime/externs.go]: I am still investigating these. The official definition have very limited support for comparison/equality operators:
https://github.com/google/cel-spec/blob/master/doc/langdef.md

Having said that, I am unearthing new information through discussions. I'll most likely open separate github issues to track these once I have a reasonable plan.

**Comment by ldemailly** [comment]
this was at some point failing the e2e mixer test, did you look at the failure in the link above ?

**Comment by ozevren** [comment]
Yes, I am looking at them.

**Pull request merged**